test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,461 @@
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_TEST_BACKWARD = Template("""\n
+void ${module_variant_name}_test_backward(
+    const std::string& saved_module_path,
+    const std::string& saved_grad_module_path,
+    const std::string& device,
+    ${input_arg_declarations}) {
+  ${module_qualified_name} grad_module${cpp_constructor_args};
+  torch::load(grad_module, saved_grad_module_path);
+
+  torch::manual_seed(2);
+  ${module_qualified_name} module${cpp_constructor_args};
+  torch::load(module, saved_module_path);
+  module->to(device);
+
+  auto cpp_output = module(${input_args});
+  cpp_output.sum().backward();
+
+  for (size_t i = 0; i < module->parameters().size(); i++) {
+    auto named_param = module->named_parameters()[i];
+    auto grad = grad_module->parameters()[i];
+    TORCH_CHECK(
+      named_param->grad().allclose(grad),
+      parity_test_error_msg, ": ", "gradient value of `", named_param.key(), "` doesn't match");
+  }
+
+  ${extra_stmts}
+}
+""")
+
+TORCH_NN_MODULE_IGNORED_ATTRS = {
+    '_backend', '_parameters', '_buffers', '_backward_hooks', '_forward_hooks', '_forward_pre_hooks',
+    '_state_dict_hooks', '_load_state_dict_pre_hooks', '_modules', 'training', 'has_parity',
+}
+
+class TestCppApiParity(common.TestCase):
+    def _python_arg_to_cpp_arg(self, python_arg):
+        if type(python_arg) == int:
+            return 'int64_t', str(python_arg)
+        elif type(python_arg) == float:
+            return 'double', str(python_arg)
+        elif type(python_arg) == bool:
+            return 'bool', str(python_arg).lower()
+        elif type(python_arg) == str:
+            return 'std::string', '"{}"'.format(python_arg)
+        elif type(python_arg) == torch.Tensor:
+            return ('torch::Tensor',
+                    'torch::empty({})'.format(str(list(python_arg.shape)).replace('[', '{').replace(']', '}')))
+        else:
+            raise RuntimeError(
+                "{} is not a supported arg type for C++ module methods".format(type(python_default_value)))
+
+    def _compile_cpp_code_inline(self, name, cpp_sources, functions):
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=name,
+            cpp_sources=cpp_sources,
+            functions=functions,
+            verbose=False,
+        )
+        return cpp_module
+
+    # This tests that Python and C++ torch.nn modules have matching constructor arg names and types.
+    def _test_torch_nn_module_ctor_args(self, module_name):
+        python_module_class = getattr(torch.nn, module_name)
+        module_metadata = torch_nn_modules.module_metadata_map[module_name]
+        python_default_constructor_args = module_metadata['python_default_constructor_args']
+        cpp_default_constructor_args = module_metadata['cpp_default_constructor_args']
+
+        cpp_module_option = 'torch::nn::' + module_name + 'Options' + cpp_default_constructor_args
```

Comment:
nit: It's easier to read `'torch::nn::{module_name}Options{default_args}'.format(...)` than adding together some strings

Reply:
Improved